### PR TITLE
[hail] Refactor ExtractIntervalFilters

### PR DESCRIPTION
**Pull request by tpoterba**
This change is necessary to make it easy to support multiple kinds
of key patterns. In particular, I want to extract comparisons with
the key struct (or a prefix of the key struct).

### CODE REVIEW

File: hail/src/main/scala/is/hail/expr/ir/ExtractIntervalFilters.scala [hunk 1]
```diff
@@ -1,11 +1,32 @@
...
 
+import is.hail.annotations.IntervalEndpointOrdering
 import is.hail.expr.types.virtual._
 import is.hail.utils.{FastSeq, Interval, IntervalEndpoint, _}
 import is.hail.variant.{Locus, ReferenceGenome}
 import org.apache.spark.sql.Row
 
 object ExtractIntervalFilters {
+
+  case class ExtractionState(rowRef: Ref, keyFields: IndexedSeq[String]) {
```
Comment (patrick-schultz) [inline hail/src/main/scala/is/hail/expr/ir/ExtractIntervalFilters.scala]: This class seems to be assuming that `keyFields` is a prefix of the list of fields of `rowRef.typ`. Is that correct?

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/ExtractIntervalFilters.scala]: no -- they're fields that appear anywhere in rowRef.typ, and in any order.

The isKeyStructPrefix rule is wrong (rowType should be rowKeyType), I fix this in the stacked PR. I can just remove that method here.

Comment (patrick-schultz) [reply hail/src/main/scala/is/hail/expr/ir/ExtractIntervalFilters.scala]: Then you also need
```
val firstKeyType: Type = rowKeyType.types.head
def isFirstKey(ir: IR): Boolean = ir == GetField(rowRef, rowKeyType.fieldNames.head)
```

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/ExtractIntervalFilters.scala]: yeah, saw that too. Will add a test that catches it.

**Comment by tpoterba** [comment]
stacked PR merged

**Pull request closed without merging**